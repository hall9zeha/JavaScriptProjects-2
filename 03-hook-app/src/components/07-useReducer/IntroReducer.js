//Creamos un lista de tareas con un objeto
import React from 'react'

export const IntroReducer = () => {

const initialObjectState = [{
    id:123,
    todo:'any task',
    done:false
}]

//Creamos nuestra función reducer que manejará los casos de uso
//Por convención una función Reducer debe tener dos argumentos, un estado y una acción
const todoReducer = (state = initialObjectState, action)=>{
    //debe retornar un nuevo valor u objeto y no modificar el valor u objeto del estado inicial, este debe ser inmutable
    if(action?.type ==='add'){//evalueamos que tipo de acción es y si esta nula o no
        return [...state,action.payLoad]//retornamos una copia del estado anterior y el nuevo objeto agregado
    }
    return state

}

//Creamos un nuevo objeto que será nuestra segunda tarea para agregar a una supuesta lista de tareas
const newTodoObject = {
    id:456,
    todo:'other task',
    done:false
}

//obtenemos el último valor devuelto por nuestro Reducer
let todos = todoReducer();

//Ahora  la acción que recibirá nuestra función Reducer será un objeto
const todoActionAdd={
    type:'add',//debe contener un tipo de acción
    payLoad:newTodoObject//Y el objeto o valor
}

//Enviamos los argumentos a nuestro Reducer
const todo = todoReducer(todos, todoActionAdd);
console.log(todo);


  return (
    <div>
        <h1>IntroReducer</h1>
        <hr/>
        <h3>Abre la consola para ver el resultado</h3>

    </div>
    
  )
}
