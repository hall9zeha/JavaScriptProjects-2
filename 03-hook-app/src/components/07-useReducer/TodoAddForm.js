import React, { Fragment } from 'react'
import { useForm } from '../../hooks/useForm'

export const TodoAddForm = ({handleTodoAdd}) => {


    const [{description}, handleInputChange,resetForm]=useForm({
        description:''
    })


    const handleSubmit = (e)=>{
        e.preventDefault();
        //Si el input box está vacío no hacer nada
        if(description.trim().length <=1){
            return;
        }
        const newTodo = {
            id:new Date().getTime(),
            desc:description,
            done:false
        };

        handleTodoAdd(newTodo);
        resetForm();
    }

  return (
    <>
         <h4>Agregar TODO</h4>
                <hr/>
                <form onSubmit={handleSubmit}>
                <input
                    type='text'
                    name='description'
                    className='form-control'
                    placeholder='Escribe una tarea'
                    autoComplete='off'
                    value={description}
                    onChange={handleInputChange}
                />
                <button type='submit' className='btn btn-block btn-outline-primary  mt-1 '>Agregar</button>
                </form>
    </>
  )
}
