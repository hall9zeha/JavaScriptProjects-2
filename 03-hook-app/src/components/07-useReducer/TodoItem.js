import React from 'react'

export const TodoItem = ({todo,index, handleDoneTask, handleDelete}) => {
  return (
    <li key={todo.id}
        className='list-group-item'
        >
        <p className={`${todo.done && 'complete'}`}
            onClick={()=>{handleDoneTask(todo.id)}}
        >{index+1 }.{todo.desc}</p>
        <button 
            data-testid='btn-delete'
            onClick={()=>{handleDelete(todo.id)}}
            className='btn btn-danger'>Borrar</button> 
    </li>
  )
}
