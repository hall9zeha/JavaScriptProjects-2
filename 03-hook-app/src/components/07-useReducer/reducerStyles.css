body{
    padding: 50px;
}
p{
    margin: 0px;
}
.list-group-item{
    align-items: center;
    display: flex;
    justify-content: space-between;
}
.complete{
    text-decoration:line-through;
}
.btn-block{
    display: block;
    width: 100%;
}