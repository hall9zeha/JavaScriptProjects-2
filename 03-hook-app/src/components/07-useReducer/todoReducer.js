

export const todoReducer = (state = [], action ) => {
  switch (action.type) {
    case 'add':
        return [...state,action.payload];//ya que retornamos directamente no necesitamos  break
        
    default:
        return state;
  }
}
