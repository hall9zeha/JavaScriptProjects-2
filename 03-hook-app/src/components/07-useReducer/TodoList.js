import React from 'react'
import { TodoItem } from './TodoItem'

export const TodoList = ({todos, handleDoneTask, handleDelete}) => {
  return (
    
             <ul className='list-group list-group-flush'>
                {
                    todos.map((todo, i)=>(
                      <TodoItem
                        key={todo.id}
                        todo={todo} 
                        index={i} 
                        handleDoneTask={handleDoneTask} 
                        handleDelete={handleDelete}/>
                    ))
                }
                </ul>
   
  )
}
