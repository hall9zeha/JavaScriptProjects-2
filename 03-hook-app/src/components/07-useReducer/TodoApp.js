import React, { useEffect, useReducer } from 'react'

import './reducerStyles.css'
import { TodoAddForm } from './TodoAddForm'
import { TodoList } from './TodoList'
import { todoReducer } from './todoReducer'

const init=()=>{
    // return [{
    //     id:new Date().getTime(),
    //     desc:'learning Android multiplatform',
    //     done:false
    // }]
    //Usaremos lo que almacenemos en Local storage
    return JSON.parse(localStorage.getItem('todos')) || []; //si existen retornar, sino un arreglo vacío
}
export const TodoApp = () => {

    const [todos, dispatch] = useReducer(todoReducer, [],init)
    


    //useEffect para lanzar la orden de grabar en Local storage
    useEffect(()=>{

        localStorage.setItem('todos', JSON.stringify(todos))
    },[todos]);//cuando cambie la lista de tareas se guardará una nueva versión en LocalStorage

    const handleTodoAdd = (newTodo)=>{
        dispatch({
            type:'add',
            payload:newTodo
            }
        )
    }

    //Para eliminar un registro
    const handleDelete = (todoId)=>{
        console.log(todoId);
        const actionDelete = {
            type:'delete',
            payload:todoId
        }
        dispatch(actionDelete);

    }
    const handleDoneTask =(todoId)=>{
        const actionDone={
            type:'done',
            payload:todoId
        }
        dispatch(actionDone);
    }

  return (
    <div>
        <h1>TodoApp ({todos.length})</h1>
        <hr/>
        <div className='row'>
            <div className='col-7'>
                <TodoList 
                    todos={todos}
                    handleDoneTask={handleDoneTask}
                    handleDelete={handleDelete}/>
                
            </div>
            <div className='col-5'>
               <TodoAddForm handleTodoAdd={handleTodoAdd} />
            </div>
        </div>
        
    </div>
  )
}
