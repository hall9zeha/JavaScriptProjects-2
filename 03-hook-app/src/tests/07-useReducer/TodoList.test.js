import { render} from '@testing-library/react';
import React from 'react';

import renderer from "react-test-renderer";
import { TodoList } from '../../components/07-useReducer/TodoList';
import { todos } from '../../fixtures/DemoTodos';

const handleDeleteMocked = jest.fn();
const handleDoneTaskMocked = jest.fn();

describe('Tests in <TodoList/> component', () => { 
    test('should show successful', () => { 
        const wrapper = renderer.create(<TodoList 
            todos={todos}
            handleDoneTask={handleDoneTaskMocked} 
            handleDelete={handleDeleteMocked}/>).toJSON();

        expect(wrapper).toMatchSnapshot()

     })
     test('should show the elements number of TODOS send',()=>{
        const   {container} = render(<TodoList 
            todos={todos}
            handleDoneTask={handleDoneTaskMocked} 
            handleDelete={handleDeleteMocked}/>);
                    

        const todoListItems = container.getElementsByClassName('list-group-item');
        expect(todoListItems.length).toBe(todos.length);
       
        
     })

 })