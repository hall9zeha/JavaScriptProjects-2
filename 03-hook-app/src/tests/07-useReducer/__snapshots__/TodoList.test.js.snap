// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <TodoList/> component should show successful 1`] = `
<ul
  className="list-group list-group-flush"
>
  <li
    className="list-group-item"
  >
    <p
      className="false"
      onClick={[Function]}
    >
      1
      .
      learning Android multiplatform
    </p>
    <button
      className="btn btn-danger"
      data-testid="btn-delete"
      onClick={[Function]}
    >
      Borrar
    </button>
  </li>
  <li
    className="list-group-item"
  >
    <p
      className="false"
      onClick={[Function]}
    >
      2
      .
      learning Angular
    </p>
    <button
      className="btn btn-danger"
      data-testid="btn-delete"
      onClick={[Function]}
    >
      Borrar
    </button>
  </li>
</ul>
`;
